packages/wonder-blocks-data/components/data.js

```diff
@@ -0,0 +1,187 @@
+    ): boolean {
+        /**
+         * We only bother updating if our state changed.
+         *
+         * And we only update the state if props changed
+         * or we got new data/error.
+         */
+        if (!this._propsMatch(nextProps)) {
+            const newState = this._buildStateAndfulfillNeeds();
+            this.setState(newState);
+        }
+
+        return (
+            this.state.loading !== nextState.loading ||
+            this.state.data !== nextState.data ||
+            this.state.error !== nextState.error
+        );
+    }
+
+    componentWillUnmount() {
+        this._mounted = false;
+    }
+
+    _propsMatch(otherProps: $Shape<Props<TOptions, TData>>): boolean {
+        const {handler, options} = this.props;
+        const {handler: prevHandler, options: prevOptions} = otherProps;
+        return (
+            handler === prevHandler &&
+            handler.getKey(options) === prevHandler.getKey(prevOptions)
+        );
+    }
+
+    _buildStateAndfulfillNeeds(): State<TData> {
+        const propsAtFulfillment = this.props;
+        const {handler, options} = propsAtFulfillment;
+        const {getEntry} = ResponseCache.Default;
+
+        const cachedData = getEntry(handler, options);
+        if (!Server.isServerSide() && cachedData == null) {
+            /**
+             * We're not on the server and the cache missed, or the cache is
+             * invalid.
+             * Therefore, we need to request data.
+             *
+             * We have to do this here from the constructor so that this
+             * data request is tracked when performing server-side rendering.
+             */
+            RequestFulfillment.Default.fulfill(handler, options)
+                .then((cacheEntry) => {
+                    /**
+                     * We get here, we should have updated the cache.
+                     * However, we need to update the component, but we
+                     * should only do that if the props are the same as they
+                     * were when this was called.
+                     */
+                    if (this._mounted && this._propsMatch(propsAtFulfillment)) {
+                        this.setState({
+                            loading: false,
+                            data: cacheEntry.data,
+                            error: cacheEntry.error,
+                        });
+                    }
+                    return null;
+                })
+                .catch((e) => {
+                    /**
+                     * We should never get here, but if we do.
+                     */
+                    if (this._mounted && this._propsMatch(propsAtFulfillment)) {
+                        this.setState({
+                            loading: false,
+                            data: null,
+                            error: typeof e === "string" ? e : e.message,
+                        });
+                    }
+                    return null;
```

Comment:
This is where we should probably log something since we really should never see this.